lib/middleware-common/src/metering.rs

```diff
@@ -1,37 +1,118 @@
-/// Metering is a compiler middleware that calculates the cost of WebAssembly instructions at compile
-/// time and will count the cost of executed instructions at runtime. Within the Metering functionality,
-/// this instruction cost is called `points`.
+/// Metering is a compiler middleware that calculates the cost of WebAssembly instructions at
+/// compile time and will count the cost of executed instructions at runtime. Within the Metering
+/// functionality, this instruction cost is called `points`.
 ///
-/// The Metering struct takes a `limit` parameter which is the maximum number of points which can be
-/// used by an instance during a function call. If this limit is exceeded, the function call will
-/// trap. Each instance has a `points_used` field which can be used to track points used during
-/// a function call and should be set back to zero after a function call.
+/// Each instance has an `exec_limit` which is the maximum number of points which can be used by
+/// the instance during a function call. If this limit is exceeded, the function call will trap.
+/// Each instance has a `points_used` field which can be used to track points used during a
+/// function call and should be set back to zero after a function call.
 ///
 /// Each compiler backend with Metering enabled should produce the same cost used at runtime for
 /// the same function calls so we can say that the metering is deterministic.
-///
 pub struct Metering {
-    limit: u64,
-    current_block: u64,
+    injections: Vec<Injection>,
+    current_block_injections: Vec<Injection>,
+    current_block_cost: u64,
 }
 
 impl Metering {
-    pub fn new(limit: u64) -> Metering {
+    pub fn new() -> Metering {
         Metering {
-            limit,
-            current_block: 0,
+            injections: Vec::new(),
+            current_block_injections: Vec::new(),
+            current_block_cost: 0,
+        }
+    }
+
+    fn set_costs<'a, 'b: 'a>(&mut self) {
+        for inj in &mut self.current_block_injections {
+            inj.check += self.current_block_cost;
+        }
+        // Set add of the last injection
+        if self.current_block_injections.len() > 0 {
+            let last_idx = self.current_block_injections.len() - 1;
+            self.current_block_injections[last_idx] = Injection {
+                add: self.current_block_cost,
+                check: 0,
+            };
+        }
+        self.current_block_cost = 0;
+    }
+
+    fn begin<'a, 'b: 'a>(&mut self) {
+        self.set_costs();
+        self.current_block_injections
+            .push(Injection { add: 0, check: 0 });
+    }
+    fn end<'a, 'b: 'a>(&mut self) {
+        self.set_costs();
+        self.injections.append(&mut self.current_block_injections);
+    }
+
+    fn inject_metering<'a, 'b: 'a>(&self, sink: &mut EventSink<'a, 'b>) {
+        let prev: Vec<Event> = sink.buffer.drain(..).collect();
+        let mut inj_idx: usize = 1;
+        for ev in prev {
+            match ev {
```

Comment:
It looks like we could put `sink.push(ev);` here and remove it everywhere else.  Feel free to not do that if you think that there will be branches in the future where that's not the case

Reply:
In some cases the event still needs to be known after it is pushed, which would move the event and make it unusable. So I opted for being explicit and putting `sink.push(ev)` every place it is needed over trying to find a more clever way to reduce code.

Reply:
Oh that's fair -- if you need it in the branches, then you can probably do it at the end of the loop 🤔 ; yeah, I was just thinking that if we want to push it every time, it's more "obviously correct" if the push happens unconditionally somewhere.  As it is now, the reader has to read all the branches to find out what the logic is in regards to the event and when editing has to make sure to do it in every new branch.  If it makes sense for there to be branches where it will not be pushed, then I think it's okay, but otherwise I think having it in one place is easier to understand.

Reply:
I'll take another look and see if I can't find a more elegant solution.

Reply:
I just took another look and I see what you're saying now, if it doesn't make sense to do this, then that's fine too!